components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt

```diff
@@ -173,29 +222,81 @@ class Nimbus(
-                nimbus.getActiveExperiments().let {
-                    if (it.any()) {
-                        recordExperimentTelemetry(it)
-                        // The current plan is to have the nimbus-sdk updateExperiments() function
-                        // return a diff of the experiments that have been received, at which point we
-                        // can emit the appropriate telemetry events and notify observers of just the
-                        // diff
-                        notifyObservers { onUpdatesApplied(it) }
-                    }
-                }
-            } catch (e: ErrorException.RequestError) {
-                logger.info("Error fetching experiments from endpoint: $e")
-            } catch (e: ErrorException.InvalidExperimentResponse) {
-                logger.info("Invalid experiment response: $e")
+        networkScope.launch {
+            fetchExperimentsOnThisThread()
+            applyPendingExperimentsOnThisThread()
+        }
+    }
+
+    override fun fetchExperiments() {
+        networkScope.launch {
+            fetchExperimentsOnThisThread()
+        }
+    }
+
+    @WorkerThread
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    private fun fetchExperimentsOnThisThread() {
+        try {
+            nimbus.fetchExperiments()
+            notifyObservers { onExperimentsFetched() }
+        } catch (e: ErrorException.RequestError) {
+            logger.info("Error fetching experiments from endpoint: $e")
+        }
+    }
+
+    override fun applyPendingExperiments() {
+        dbScope.launch {
+            applyPendingExperimentsOnThisThread()
+        }
+    }
+
+    @WorkerThread
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    private fun applyPendingExperimentsOnThisThread() {
+        try {
+            nimbus.applyPendingExperiments()
+            // Get the experiments to record in telemetry
+            postEnrolmentCalculation()
+        } catch (e: ErrorException.InvalidExperimentFormat) {
+            logger.info("Invalid experiment format: $e")
+        }
+    }
+
+    private fun postEnrolmentCalculation() {
+        nimbus.getActiveExperiments().let {
+            if (it.any()) {
+                recordExperimentTelemetry(it)
+                // The current plan is to have the nimbus-sdk updateExperiments() function
+                // return a diff of the experiments that have been received, at which point we
+                // can emit the appropriate telemetry events and notify observers of just the
+                // diff
+                notifyObservers { onUpdatesApplied(it) }
+            }
+        }
+    }
+
+    override fun setExperimentsLocally(@RawRes file: Int) {
+        dbScope.launch {
+            val payload = context.resources.openRawResource(file).use {
+                it.bufferedReader().readText()
             }
+            setExperimentsLocallyOnThisThread(payload)
```

Comment:
For parity with `fetch` we should probably put this on the network thread.

Reply:
Hmm. On second thoughts, this is probably not desirable to have it on the network thread: it's use case is to be called immediately before an `apply` so it would be a self-own to set up a race between the `fetch` and `db` threads.

Reply:
What are the semantics of the "network thread"? I don't think there's a single call we have which does network but no disk IO, so if the desired semantics of the network thread is "doesn't do disk IO", it might not be appropriate here.

Reply:
No, you're right. I renamed the `networkScope` the `fetchScope`. It is only for the slow calls out to the network and the stashing of what is found there (i.e. the everything else uses the `dbScope`).